crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+
+    for (i = 0; i != nkdfs; i++) {
+        if (hpke_kdf_tab[i].kdf_id == codepoint) {
+            return (i);
+        }
+    }
+    return (0);
+}
+
+/*
+ * @brief Check if kem_id is ok/known to us
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for good, not 1 for error
+ */
+static int hpke_kem_id_check(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
+    case OSSL_HPKE_KEM_ID_P384:
+    case OSSL_HPKE_KEM_ID_P521:
+    case OSSL_HPKE_KEM_ID_25519:
+    case OSSL_HPKE_KEM_ID_448:
+        break;
+    default:
+        return (- __LINE__);
+    }
+    return (1);
+}
+
+/*
+ * @brief check if KEM uses NIST curve or not
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for NIST, 0 for good-but-non-NIST, other otherwise
+ */
+static int hpke_kem_id_nist_curve(uint16_t kem_id)
+{
+    if (hpke_kem_id_check(kem_id) != 1)
+        return (- __LINE__);
+    if (kem_id >= 0x10 && kem_id < 0x20)
+        return (1);
+    return (0);
+}
+
+/*
+ * @brief hpke wrapper to import NIST curve public key as easily as x25519/x448
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param curve is the curve NID
+ * @param gname is the curve groupname
+ * @param buf is the binary buffer with the (uncompressed) public value
+ * @param buflen is the length of the private key buffer
+ * @return a working EVP_PKEY * or NULL
+ */
+static EVP_PKEY * hpke_EVP_PKEY_new_raw_nist_public_key(OSSL_LIB_CTX *libctx,
+                                                        int curve,
+                                                        const char *gname,
+                                                        unsigned char *buf,
+                                                        size_t buflen)
+{
+    int erv = 1;
+    EVP_PKEY *ret = NULL;
+    EVP_PKEY_CTX *cctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", NULL);
+
+    if (cctx == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_paramgen_init(cctx) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(cctx, curve) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_paramgen(cctx, &ret) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_set1_encoded_public_key(ret, buf, buflen) != 1) {
+        EVP_PKEY_free(ret);
+        ret = NULL;
+        OSSL_HPKE_err;
+        goto err;
+    }
+
+err:
+    EVP_PKEY_CTX_free(cctx);
+    if (erv == 1)
+        return (ret);
+    else
+        return NULL;
+}
+
+/*
+ * There's an odd accidental coding style feature here:
+ * For all the externally visible functions in hpke.h, when
+ * passing in a buffer, the length parameter precedes the
+ * associated buffer pointer. It turns out that, entirely by
+ * accident, I did the exact opposite for all the static
+ * functions defined inside here. But since I was consistent
+ * in both cases, I'll declare that a feature and move on:-)
```

Comment:
OpenSSL coding style is to put the buffer first followed by the length. We should do that everywhere.

Reply:
Yep, fixed due to comment above.